### PR TITLE
Handle unexpected response status in async methods

### PR DESCRIPTION
**Pull request by RikkiGibson**
I wound up factoring out a few helpers and doing some slightly odd-looking things to make this work.

A particular difficult point was to implement handleSyncHttpResponse in terms of handleAsyncHttpResponse, because handleAsyncHttpResponse can return either Completable or Single, and handleAsyncHttpResponse makes incompatible assumptions about the return type of the Swagger method. I expect a little revision and maybe some more test coverage to be necessary before merge.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
It looks like some tests failed in CI, but they're passing on my local machine. I can't restart the build, and I'm not sure what the failure is about.

**Comment by daschult** [comment]
Those failures are about the request/response body is a String, should it be quoted or not? I felt that it shouldn't be quoted, so I made some tests to enforce that.

**Comment by RikkiGibson** [comment]
I was able to repro the error locally after merging. I just can't see what changed about the way I get the response content and deserialize it that would've caused the quotes go to away.

This test is deserializing to HttpBinJSON and printing the .data property.

File: client-runtime/src/main/java/com/microsoft/rest/RestProxy.java [hunk 1]
```diff
@@ -165,84 +166,99 @@ protected Object handleSyncHttpResponse(HttpRequest httpRequest, HttpResponse ht
         return handleSyncHttpResponse(httpRequest, httpResponse, methodParser, returnType);
     }
 
-    protected Object handleSyncHttpResponse(HttpRequest httpRequest, HttpResponse httpResponse, SwaggerMethodParser methodParser, Type returnType) throws IOException {
-        Object result;
+    protected Object handleSyncHttpResponse(HttpRequest httpRequest, HttpResponse httpResponse, SwaggerMethodParser methodParser, final Type returnType) throws IOException {
```
Comment (daschult) [inline client-runtime/src/main/java/com/microsoft/rest/RestProxy.java]: `returnType` doesn't need to be `final` here anymore.

**Pull request merged**